### DISCUSSION TITLE
Should I expect relations to be included in GQL results automatically when using this library?

### DISCUSSION THREAD

**Discussion by abrenoch** [TriPSs/nestjs-query#226] [Q&A] [answered]
Hi, a bit new to nestjs and this library.

I am having a rough time getting my relationships to be included in my gql query results. I have gone through the docs several times and it is not clear to me if `nestjs-query` should be including these relations automatically or if a dataloader (or something like that) needs to be used.

**assets.entity.ts**
```ts
@Entity({ name: 'assets' })
export class AssetEntity {

  @PrimaryGeneratedColumn()
  id: number;

  @Column({ nullable: true, name: 'file_id' })
  fileId: number;

  @Column({ name: 'asset_name' })
  assetName: string;

  @ManyToOne(() => FileEntity, (fileEntity) => fileEntity.assets)
  @JoinColumn({ name: 'file_id' })
  file?: FileEntity;

  @OneToMany(() => AssetCategoriesEntity, (entity) => entity.asset)
  @JoinColumn({ name: 'asset_id' })
  categories: AssetCategoriesEntity[];
}
```

**assets.model.ts**
```ts
@ObjectType('Asset', {
  description: 'Asset model',
})
@UnPagedRelation('categories', () => AssetCategoriesModel, {
  nullable: true,
})
@Relation('file', () => FileModel, {
  nullable: true,
  enableLookAhead: true,
})
@InputType()
export class AssetModel {

  @IDField(() => Int)
  id!: number;

  @Field(() => Int, { nullable: true })
  fileId: number;

  @Field({ nullable: true })
  assetName: string;
}
```

**asset-categories.entity.ts**
```ts
@Entity({ name: 'asset_categories' })
export class AssetCategoriesEntity {
  @PrimaryGeneratedColumn()
  id: number;

  @Column({ name: 'asset_id' })
  assetId: number;

  @ManyToOne(
    () => AssetEntity,
    (entity) => entity.categories,
  )
  @JoinColumn([{ name: 'asset_id' }])
  asset: AssetEntity;
}
```

**asset-categories.model.ts**
```ts
@ObjectType('AssetCategories', {
  description: 'Asset Categories model',
})
export class AssetCategoriesModel {
  @IDField(() => ID)
  id!: number;
}
```

I can see the query for the categories actually being used:

```sql
SELECT DISTINCT `categories`.`id` AS `categories_id`, `categories`.`asset_id` AS `categories_asset_id` FROM `asset_categories` `categories` WHERE `categories`.`asset_id` IN (?)
```
But nothing is returned:

```
query {
  getAsset(
    id: 1127728
  ) {
    id
    assetName
    categories {
      __typename
      id
    }
  }
}
======
{
  "data": {
    "getAsset": {
      "id": 1127728,
      "assetName": "<asset_name>",
      "categories": []
    }
  }
}
```

*Clearly* I'm missing something; I'll happily provide more details if it would help.

Should I be making custom dataloaders for these? I just wanted to clarify before going down that path.

Any help is appreciated, even if it to point out something obvious I missed 😄

**Comment by TriPSs** [comment]
Everything looks ok, stupid questions: 
1. That asset does have categories?
2. If you run the query yourself does it than return data?

And could you try removing the `@JoinColumn({ name: 'asset_id' })` inside the `AssetEntity`, the `JoinColumn` should only be applied to one side of the relation.

**Reply by abrenoch** [reply]
Hey thanks for the quick response!

1 - It does indeed, I just tried one with 2 categories and none were returned.
2 - Can confirm 2 items are returned when testing the DB query typeorm spits to the console.

I also tried removing the join column (I tried in both the `AssetEntity` and `AssetCategoriesEntity`) but didn't seem to help.

Is there anything I can provide that might give more clues? I can look into setting up a minimal test repo if that might help.

**Reply by TriPSs** [reply]
Could you give it a try with updating the `asset-categories.entity.ts` to:
```ts
@Entity({ name: 'asset_categories' })
export class AssetCategoriesEntity {
  @PrimaryGeneratedColumn()
  id: number;

  @Column({ name: 'asset_id' })
  assetId: number;

  @ManyToOne(() => AssetEntity)
  @JoinColumn([{ name: 'asset_id' }])
  asset: AssetEntity;
}
```

**Reply by abrenoch** [reply]
Unfortunately that didnt seem to help 😞 - I'll start putting a little test repo together

**Reply by TriPSs** [reply]
Maybe it's related to #89, example repo would be great!

Could you maybe give this a try, just to test?
```ts
@Entity({ name: 'asset_categories' })
export class AssetCategoriesEntity {
  @PrimaryGeneratedColumn()
  id: number;

  @Column()
  asset_id: number;

  @ManyToOne(() => AssetEntity)
  @JoinColumn([{ name: 'asset_id' }])
  asset: AssetEntity;
}
```

**Reply by abrenoch** [reply]
I'll take a look at that issue - I've tried some different values and renaming the relation but that didnt get me anywhere.

I did notice my `@JoinColumn` is passed an array rather than an object, which gave me a glimmer of hope that was quickly dashed away 
when changing it didnt fix anything 😄.

**Reply by abrenoch** [reply]
@TriPSs here is a *minimal* example still producing the same issue for me: https://github.com/abrenoch/nest-query-test

Any insight or feedback is greatly appreciated!

**Reply by TriPSs** [reply, marked as answer]
Thanks for the repo, I have found the issue: The mapper expects all the types to be correct, your database is using `bigint` which causes the `rawRelations` to look like this:
```
[
  { categories_id: '1', categories_asset_id: '1' },
  { categories_id: '2', categories_asset_id: '1' }
]
```

As you can see all the id's are now string, if we change the big int to integer the raw relations look like this:
```
[
  { categories_id: 1, categories_asset_id: 1 },
  { categories_id: 2, categories_asset_id: 1 }
]
```
Now all the id's are numbers, which is can than correctly be mapped back to the real entities values.

So, long story short, you can disable this behaviour by setting `bigNumberStrings` in your TypeORM config to `false`.

**Reply by abrenoch** [reply]
@TriPSs My hero! Where do I send the tip?

Thanks a **ton**, I wouldn't have suspected a bigint would cast to a string. Makes sense now!
